kanidmd/src/lib/filter.rs

```diff
@@ -1131,31 +1187,25 @@ impl FilterResolved {
                 // AndNot currently is a rare request, cloning is not the worst thing
                 // here ...
                 FilterResolved::resolve_no_idx((*f).clone(), ev)
-                    .map(|fi| FilterResolved::AndNot(Box::new(fi)))
+                    .map(|fi| FilterResolved::AndNot(Box::new(fi), None))
             }
-
-            FilterComp::SelfUuid => ev.get_uuid().map(|uuid| {
-                FilterResolved::Eq(
-                    AttrString::from("uuid"),
-                    PartialValue::new_uuid(uuid),
-                    false,
-                )
-            }),
         }
     }
 
     // This is an optimise that only attempts to optimise the outer terms.
     fn fast_optimise(self) -> Self {
         match self {
-            FilterResolved::Inclusion(mut f_list) => {
+            FilterResolved::Inclusion(mut f_list, _) => {
                 f_list.sort_unstable();
                 f_list.dedup();
-                FilterResolved::Inclusion(f_list)
+                let sf = f_list.last().and_then(|f| f.get_slopeyness_factor());
```

Comment:
Why do we need the slopeyness factor of the "greatest" element? How are these even being compared when `sort_unstable` is called?

Reply:
An inclusion is similar to an OR - so we want to know "what's the most expensive part of this query going to be" since we have to access EVERY index in the OR/Inclusion term. This way if we have a query like:

`x or inclusion(a, b, c)`

we want to check x first since it may *fail* faster allowing us to skip the more expensive inclusion. If we get to the expensive inclusion term, then we want to be at a point where our partial candidate set is smaller, allowing quicker merging of the data, and knowing that we will need to actually process it.

Reply:
Okay, the "what's the most expensive part of this query going to be" was the part that I wasn't sure about, but definitely aligns with the code. Also, what's the difference between `Inclusion` and `Or`?

Reply:
Inclusion are an internal piece of magic for plugin behaviour - they can't be generated from external calls, as they don't "mean" anything in a search query. Specifically it's used in referential integrity https://github.com/kanidm/kanidm/blob/master/kanidmd/src/lib/plugins/refint.rs#L60 . The server has a second "lighter" search path for "existance" or rather "does this query return true/false". Inclusion is different to AND/OR because:

* AND expresses that all terms must hold true.
* OR expresses that at least one term holds true.
* Inclusion expresses that every term holds true at least ONCE.

Consider a query like:

```
uuid=A  OP uuid=B
```

If OP was AND, then this would always be false - we don't have an entry that could have both uuid's, since they are on seperate entries.

If OP was OR, then this would be true, provided A OR B exist.

But in referential integrity we need assert uuid=A and uuid=B both exist, but may be on seperate entries. That's what inclusion does. It allows expressing that each term must hold true on at least one entry, and any term that is false, will shortcut us out. There are some extra rules in inclusion/exists queries (they don't have a filter threshold because we need to force every indexed element of the inclusion to be evaluated else we'll have an inconsistent candidate set, exists queries disable shortcutting because we want to test indexes as much as possible as we wont load entries etc.)

The point is - Inclusion allows refint to assert the existance of many things in one query, rather than issuing many individual queries in sequence, and the fn exists allows testing a query for truth without loading entries from id2entry.